src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicSession.cs

```diff
@@ -0,0 +1,155 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Security;
+using System.Text;
+
+namespace System.Net.Quic.Implementations.MsQuic.Internal
+{
+    internal sealed class MsQuicSession : IDisposable
+    {
+        private bool _disposed = false;
+        private IntPtr _nativeObjPtr;
+        private MsQuicApi _api;
+        private bool _opened;
+
+        internal MsQuicSession(string registration)
+        {
+            _api = new MsQuicApi();
+
+            _api.RegistrationOpen(Encoding.UTF8.GetBytes(registration));
+        }
+
+        public MsQuicConnection ConnectionOpen(IPEndPoint endpoint, SslClientAuthenticationOptions sslClientAuthenticationOptions, IPEndPoint localEndpoint)
+        {
+            if (!_opened)
+            {
+                _nativeObjPtr = _api.SessionOpen(sslClientAuthenticationOptions.ApplicationProtocols[0].Protocol.ToArray());
+                _opened = true;
+                SetPeerBiDirectionalStreamCount(100);
+                SetPeerUnidirectionalStreamCount(100);
+            }
+
+            uint status = _api._connectionOpenDelegate(
+                _nativeObjPtr,
+                MsQuicConnection.NativeCallbackHandler,
+                IntPtr.Zero,
+                out IntPtr connectionPtr);
+
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            return new MsQuicConnection(endpoint, _api, connectionPtr, sslClientAuthenticationOptions);
+        }
+
+        public MsQuicListener ListenerOpen(IPEndPoint listenEndPoint, SslServerAuthenticationOptions sslServerAuthenticationOptions)
+        {
+            if (!_opened)
+            {
+                _nativeObjPtr = _api.SessionOpen(sslServerAuthenticationOptions.ApplicationProtocols[0].Protocol.ToArray());
+                _opened = true;
+                // TODO figure out a story for configuration.
+                SetPeerBiDirectionalStreamCount(100);
+                SetPeerUnidirectionalStreamCount(100);
+            }
+
+            uint status = _api._listenerOpenDelegate(
+                _nativeObjPtr,
+                MsQuicListener.NativeCallbackHandler,
+                IntPtr.Zero,
+                out IntPtr listenerPointer
+                );
+
+            var listener = new MsQuicListener(listenEndPoint, sslServerAuthenticationOptions, _api, listenerPointer);
+
+            return listener;
+        }
+
+        public void ShutDown(
+            QUIC_CONNECTION_SHUTDOWN_FLAG Flags,
+            ushort ErrorCode)
+        {
+            _api._sessionShutdownDelegate(
+                _nativeObjPtr,
+                (uint)Flags,
+                ErrorCode);
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        public void SetPeerBiDirectionalStreamCount(ushort count)
```

Comment:
I think I mentioned something similar on the connection or stream, but I think it makes sense to have QUIC parameters be C# properties.